utils/merge.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base or patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - base and patch must have no unexported types (at the moment)
+//   - if maps or slices are different, the entire map or slice will be replaced (at the moment)
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//	   ret, err := merge(base, patch)
+//	   if err != nil {
+//         return nil, err
+//	   }
+//	   retTS := ret.(testStruct)
+//	   return &retTS, nil
+// }
+func Merge(base interface{}, patch interface{}) (interface{}, error) {
+	if reflect.TypeOf(base) != reflect.TypeOf(patch) {
+		return nil, fmt.Errorf("cannot merge different types. base type: %s, patch type: %s",
+			reflect.TypeOf(base), reflect.TypeOf(patch))
+	}
+
+	baseType := reflect.TypeOf(base)
+	if baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+	}
+
+	ret := reflect.New(baseType)
+	if err := initStruct(baseType, ret.Elem()); err != nil {
+		return nil, err
+	}
+
+	if err := mergeRec(base, patch, ret.Elem()); err != nil {
+		return nil, err
+	}
+	return ret.Elem().Interface(), nil
+}
+
+// mergeRec recursively merges into ret
+func mergeRec(base interface{}, patch interface{}, ret reflect.Value) (err error) {
```

Comment:
Similar comment to the other review, but I feel like the following signature would be more natural for a recursive method:
```
func mergeRec(base interface{}, patch interface{}) (ret reflect.Value, error)
```